File: runtime/runtime.go
/*
	Copyright 2022 Loophole Labs

	Licensed under the Apache License, Version 2.0 (the "License");
	you may not use this file except in compliance with the License.
	You may obtain a copy of the License at

		   http://www.apache.org/licenses/LICENSE-2.0

	Unless required by applicable law or agreed to in writing, software
	distributed under the License is distributed on an "AS IS" BASIS,
	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
	See the License for the specific language governing permissions and
	limitations under the License.
*/

// Package runtime implements the Scale Runtime in Go.
package runtime

import (
	"context"
	"fmt"
	"github.com/google/uuid"
	"github.com/loopholelabs/scale-go/scalefunc"
	"github.com/tetratelabs/wazero"
	"github.com/tetratelabs/wazero/api"
)

// Next is the next function in the middleware chain. It's meant to be implemented
// by whatever adapter is being used.
type Next func(ctx *Context) *Context

// Runtime is the Scale Runtime. It is responsible for initializing
// and managing the WASM runtime as well as the scale function chain.
type Runtime struct {
	runtime       wazero.Runtime
	compileConfig wazero.CompileConfig
	moduleConfig  wazero.ModuleConfig

	functions []*Function
	instances map[string]*Instance
}

func New(ctx context.Context, functions []scalefunc.ScaleFunc) (*Runtime, error) {
	r := &Runtime{
		runtime:       wazero.NewRuntimeWithConfig(wazero.NewRuntimeConfig().WithWasmCore2()),
		compileConfig: wazero.NewCompileConfig(),
		moduleConfig:  wazero.NewModuleConfig(),
		instances:     make(map[string]*Instance),
	}

	module := r.runtime.NewModuleBuilder("env")
	module = module.ExportFunction("next", r.next)
	//module = module.ExportFunction("debug", r.debug)

	compiled, err := module.Compile(ctx, r.compileConfig)
	if err != nil {
		return nil, fmt.Errorf("failed to compile module: %w", err)
	}

	_, err = r.runtime.InstantiateModule(ctx, compiled, r.moduleConfig)
	if err != nil {
		return nil, fmt.Errorf("failed to instantiate module: %w", err)
	}

	module = r.runtime.NewModuleBuilder("wasi_snapshot_preview1")
	module = module.ExportFunction("fd_write", r.fd_write)

	compiled, err = module.Compile(ctx, r.compileConfig)
	if err != nil {
		return nil, fmt.Errorf("failed to compile module: %w", err)
	}

	_, err = r.runtime.InstantiateModule(ctx, compiled, r.moduleConfig)
	if err != nil {
		return nil, fmt.Errorf("failed to instantiate module: %w", err)
	}

	for _, f := range functions {
		err = r.registerFunction(ctx, f)
		if err != nil {
			return nil, err
		}
	}

	return r, nil
}

func (r *Runtime) Instance(ctx context.Context, next Next) (*Instance, error) {
	i := &Instance{
		id:      uuid.New().String(),
		next:    next,
		runtime: r,
		modules: make(map[api.Module]*Module),
		ctx:     NewContext(),
	}

	r.instances[i.id] = i

	return i, i.initialize(ctx)
}

File: ts/src/sigruntime/runtime.ts
/*
        Copyright 2022 Loophole Labs

        Licensed under the Apache License, Version 2.0 (the "License");
        you may not use this file except in compliance with the License.
        You may obtain a copy of the License at

                   http://www.apache.org/licenses/LICENSE-2.0

        Unless required by applicable law or agreed to in writing, software
        distributed under the License is distributed on an "AS IS" BASIS,
        WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
        See the License for the specific language governing permissions and
        limitations under the License.
*/

import { Signature, SignatureFactory } from "../signature/signature";
import { ScaleFunc } from "../signature/scaleFunc";

import { SFunction } from "./sfunction";
import { Instance } from "./instance";
import { Pool } from "./pool";
import { Module } from "./module";

export interface WasiContext {
  start(instance: WebAssembly.Instance): void;
  getImportObject(): any;
}

export type NextFn<T extends Signature> = (ctx: T) => T;

export class Runtime<T extends Signature> {
  public Ready: Promise<any>;
  
  public signatureFactory: SignatureFactory<T>;
  private fns: SFunction<T>[];
  public head: undefined | SFunction<T>;
  private tail: undefined | SFunction<T>;

  private wasiBuilder: () => WasiContext;

  public modules: Map<string, Module<T>>;   // Map from unique module ID to module

  constructor(wasiBuilder: () => WasiContext, sigfac: SignatureFactory<T>, fns: ScaleFunc[]) {
    this.signatureFactory = sigfac;
    this.fns = [];
    this.modules = new Map<string, Module<T>>;

    this.wasiBuilder = wasiBuilder;

    // We compile the modules async...
    // After creating a Runtime you should then do 'await runtime.Ready' or equivalent.
    this.Ready = new Promise(async (resolve, reject) => {

      for(let i=0;i<fns.length;i++) {
        const fn = fns[i];
        const mod = await WebAssembly.compile(fn.Function as Buffer);

        const f = new SFunction<T>(fn, mod);
        f.modulePool = new Pool<T>(f, this);
        this.fns.push(f);

        if (this.head === undefined) {
          this.head = f;
        }
        if (this.tail !== undefined) {
          this.tail.next = f;
        }
        this.tail = f;
      }

      resolve(true);
    });
  }

  Instance(next: null | NextFn<T>): Instance<T> {
    return new Instance<T>(this, next);
  }

  instantiate(m: WebAssembly.Module, mod: Module<T>, i: Instance<T>): WebAssembly.Instance {
    // NB This closure captures i.
    const nextFn = ((runtimeThis: Runtime<T>): Function => {
      return (ptr: number, len: number): BigInt => {
        if (mod.memory===undefined || mod.resize===undefined) {
          // Critical unrecoverable error
          // NB This would only ever happen if init() wasn't called on the Module.
          return BigInt(0);
        }

        let buff: Uint8Array = new Uint8Array();
        try {
          const memDataOut = new Uint8Array(mod.memory.buffer);
          const inContextBuff = memDataOut.slice(ptr, ptr + len);
          i.RuntimeContext().Read(inContextBuff);

          // Now call next...
          if (mod.sfunction.next === undefined) {
            i.ctx = i.next(i.Context());
            buff = i.RuntimeContext().Write();
          } else {
            mod.sfunction.next.Run(i);
            buff = i.RuntimeContext().Write();
          }
        } catch(e) {
          buff = i.RuntimeContext().Error(e as Error);          
        }

        // Write it back out
        const encPtr = mod.resize(buff.length);
        const memData = new Uint8Array(mod.memory.buffer);
        memData.set(buff, encPtr);
        return SFunction.packMemoryRef(encPtr, buff.length);
      };
    })(this);
    
    const wasi = this.wasiBuilder();
    const importObject = {
      wasi_snapshot_preview1: wasi.getImportObject(),
      env: {
        next: nextFn,
      },
    };

    // TODO: We may need to do something async for browser contexts
    // eg WebAssembly.instantiate(m, importObject);
    const inst = new WebAssembly.Instance(m, importObject);
    wasi.start(inst);
    return inst;
  }
}